metrics/prometheus-metrics-endpoints-and-auto-discovery.md

```diff
@@ -0,0 +1,136 @@
+
+For example, if a Service is exposing a HTTP metrics endpoint at `/metrics`, the annoation would be:
+
+```yaml
+org.aerogear.metrics/plain_endpoint: /metrics
+```
+
+If a Service is exposing a HTTPS metrics endpoint at `/rest/metrics`, the annoation would be:
+
+```yaml
+org.aerogear.metrics/ssl_endpoint: /rest/metrics
+```
+
+### Default Visualisations
+
+Grafana will be configured 'out of the box' with 'Dashboards' for specific Mobile Services. Which services will be determined on a case by case basis. These Dashboards will be tailored to the needs of a developer using that service. For example, the Data Sync Dashboard will show queue metrics.
+
+A Default Dashboard for Prometheus metrics will be made available. This will show, where possible, metrics that are generic across all auto-discovered services e.g. CPU/Memory.
+
+Developers will be free to add their own Dashboards, or modify/remove any default Dashboards.
+
+Default dashboards will be stored in a ConfigMap that gets mounted into the Grafana container
+e.g. https://github.com/aerogearcatalog/prometheus-apb/blob/ee86cfc67caaeaabcf7e7bc214c3a6dc6940c333/roles/provision-prometheus-apb/tasks/main.yml#L30-L33
+
+The dashboards will be picked up in Grafana by setting the path to the mounted ConfigMap resource. e.g.
+
+```ini
+[paths]
+datasources = /etc/grafana-datasources
+```
+
+### Auth
+
+Both Prometheus and Grafana will be accessible to users who have an account in the OpenShift cluster where they are deployed. This will allow for easier use of those services via single sign on.
+
+This can be achieved by using an [oauth proxy](https://github.com/openshift/oauth-proxy) in front of either service. This proxy will use an OAuth flow to authenticate the User against the OpenShift cluster. The namespace will have a ServiceAccount Oauth Client i.e. has the below annotations to allow OAuth redirects to the Grafana & Prometheus proxy routes.
+
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: proxy
+  annotations:
+    serviceaccounts.openshift.io/oauth-redirectreference.grafana: '{"kind":"OAuthRedirectReference","apiVersion":"v1","reference":{"kind":"Route","name":"grafana"}}'
+    serviceaccounts.openshift.io/oauth-redirectreference.prometheus: '{"kind":"OAuthRedirectReference","apiVersion":"v1","reference":{"kind":"Route","name":"prometheus"}}'
+```
+
+Both OAuth Proxies, Grafana & Prometheus will all run in the same Pod, but in separate containers. This is necessary so Grafana can talk directly to Promethues via localhost, bypassing the OAuth Proxy when accessing it as a Data Source. Also, the OAuth Proxies can proxy directly on localhost to Grafana & Prometheus without having to expose a Service, making the unprotected services potentially accessible outside the namespace.
```

Comment:
> Both OAuth Proxies, Grafana & Prometheus will all run in the same Pod, but in separate containers. This is necessary so Grafana can talk directly to Promethues via localhost, bypassing the OAuth Proxy when accessing it as a Data Source. 

Will this make it hard to scale the metrics service?

Reply:
"To put some numbers behind this: a single big Prometheus server can easily store millions of time series, with a record of 800,000 incoming samples per second (as measured with real production metrics data at SoundCloud). Given a 10-seconds scrape interval and 700 time series per host, this allows you to monitor over 10,000 machines from a single Prometheus server"
https://prometheus.io/blog/2016/07/23/pull-does-not-scale-or-does-it/
I say lets cross that bridge when we get to it as it will likely need a completely different approach

Reply:
@david-martin  I am not talking about how many machines a single Prometheus server can support, I am talking about having Grafana & Prometheus running in the same pod. In this setup, we can not scaling Grafana without scaling Prometheus - but as you said, a single Prometheus can support many machines, so scaling Prometheus seems unnecessary. Feels like it's better to keep them in separate pods, and figure out how to make Grafana authenticate with Prometheus?

Reply:
@wei-lee Good point. Scaling Grafana may be something the user wants to do.
From doing some reading, it should be possible to have a shared filesystem/persistent volume for grafana storoage of plugins & dashboards
https://medium.com/@fcgravalos/scaling-out-grafana-with-kubernetes-and-aws-62745257df10

Solving the auth problem between Grafana & Prometheus could be tricky. To keep the oauth-proxy in front of prometheus, grafana would need to have an oauth-proxy cookie header (which is got by logging in via oauth). I'm not sure how that could be done.

We could remove the oauth-proxy in front of prometheus, but we'd have to remove the exposed route too. Then Grafana could access it on the service name without auth i.e. 'prometheus'.

Another option could be to expose another service in front of prometheus that bypasses the oauth-proxy. This service would only be internally accessible/routable. The exposed external route to prometheus would still be routed through the oauth proxy.
However, I'm not sure if having the extra service would protect it from being accessed from another namespace via the full service host i.e. prometheus.namespace.svc.cluster.local

Reply:
Option 3 could be a runner if projects have isolated networks
https://docs.openshift.com/container-platform/3.7/admin_guide/managing_networking.html#joining-project-networks
@philbrookes any thoughts on this? I think we discussed it before but I may have ruled it out due to security concerns about accessibility from other projects. I believe that shouldn't be the case though

Reply:
For Option 1, using the `--skip-auth-regex` flag, as in the PR below could allow us to bypass the oauth proxy for certain urls
https://github.com/openshift/origin/pull/17114/files#diff-1630f01082dce4d5ef4ba69488c022a8R125

The datasource definition in the same PR uses a serviceaccount token too
https://github.com/openshift/origin/pull/17114/files#diff-cd98aab0c9f8f84ad1e48d348b459d0fR53

Reply:
Updating to go with Option 3.